python/pdo/contract/state.py

```diff
@@ -118,29 +118,47 @@ def __read_data_block_from_cache__(cls, contract_id, state_hash, data_dir = None
 
     # --------------------------------------------------
     @staticmethod
-    def __push_block_to_eservice__(eservice, contract_id, state_hash, data_dir = None) :
+    def __push_blocks_to_eservice__(eservice, contract_id, block_ids, data_dir = None) :
         """
-        ensure that a particular block is stored in the eservice
+        ensure that required blocks are stored in the storage service
 
         :param eservice EnclaveServiceClient object:
         :param contract_id str: contract identifier
-        :param state_hash string: base64 encoded hash of the block
+        :param block_ids list of strings: base64 encoded hash of the block
         """
 
-        logger.debug('ensure block %s is stored in the eservice', state_hash)
-
-        # check to see if the eservice already has the block
-        if eservice.block_store_head(state_hash) > 0 :
-            return False
-
-        raw_data = ContractState.__read_data_block_from_cache__(contract_id, state_hash, data_dir)
-        if raw_data is None :
-            raise Exception('unable to locate required block; {}'.format(state_hash))
-
-        eservice.block_store_put(state_hash, raw_data)
-
-        logger.debug('sent block %s to eservice', state_hash)
-        return True
+        # check to see which blocks need to be pushed
+        blocks_to_push = []
+        blocks_to_extend = []
+        block_status_list = eservice.check_blocks(block_ids)
+        for block_status in block_status_list :
+            # if the size is 0 then the block is unknown to the storage service
+            if block_status['size'] == 0 :
+                blocks_to_push.append(block_status['block_id'])
+            # if the expiration is nearing, then add to the list to extend, the
+            # policy here is to extend if the block is within 5 seconds of expiring
+            elif block_status['expiration'] < 5 :
+                blocks_to_extend.append(block_status['block_id'])
```

Comment:
blocks_to_extend is unused after creation. Is the extension policy implemented in a different PR?

Reply:
will fix and resubmit

Reply:
fixed